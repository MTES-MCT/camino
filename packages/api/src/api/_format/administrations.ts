import { IAdministration } from '../../types'

import { titresFormat } from './titres'
import { utilisateurFormat } from './utilisateurs'
import {
  Administrations,
  ADMINISTRATION_TYPES
} from 'camino-common/src/administrations'
import { Regions } from 'camino-common/src/region'
import { Departements } from 'camino-common/src/departement'

/**
 * Formate une administration en fonction du profil de l'utilisateur
 *
 * @param administration - Administration à formater
 * @returns Une administration formatée
 *
 */

const administrationFormat = (administration: IAdministration) => {
  administration.gestionnaireTitres =
    administration.gestionnaireTitres &&
    titresFormat(administration.gestionnaireTitres)

  administration.localeTitres =
    administration.localeTitres && titresFormat(administration.localeTitres)

  administration.utilisateurs =
    administration.utilisateurs?.map(utilisateurFormat)

  const adminis = Administrations[administration.id]
  administration.type = ADMINISTRATION_TYPES[adminis.typeId]

  for (const key of Object.keys(adminis)) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    administration[key] = adminis[key]
  }

  if (adminis.departementId) {
    administration.departement = Departements[adminis.departementId]
  }
  if (adminis.regionId) {
    administration.region = Regions[adminis.regionId]
  }

  return administration
}

export { administrationFormat }
