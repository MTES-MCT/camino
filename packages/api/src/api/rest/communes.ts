import { Commune, communeIdValidator } from 'camino-common/src/static/communes.js'
import { Pool } from 'pg'
import { HTTP_STATUS } from 'camino-common/src/http.js'
import { CaminoRequest, CustomResponse } from './express-type.js'
import { getCommunes as getCommunesQuery } from '../../database/queries/communes.queries.js'
import { z } from 'zod'

export const getCommunes = (pool: Pool) => async (req: CaminoRequest, res: CustomResponse<Commune[]>) => {
  try {
    const communeIds = z.array(communeIdValidator).nonempty().safeParse(req.query.ids)
    if (communeIds.success) {
      const communes = await getCommunesQuery(pool, { ids: communeIds.data })
      res.json(communes)
    } else {
      res.sendStatus(HTTP_STATUS.HTTP_STATUS_BAD_REQUEST)
    }
  } catch (e) {
    console.error(e)

    res.sendStatus(HTTP_STATUS.HTTP_STATUS_INTERNAL_SERVER_ERROR)
  }
}
