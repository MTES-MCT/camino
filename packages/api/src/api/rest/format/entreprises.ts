import { IEntreprise } from '../../../types.js'

const entreprisesFormatTable = (entreprises: IEntreprise[]) =>
  entreprises.map(entreprise => {
    const entrepriseNew = {
      nom: entreprise.nom,
      siren: entreprise.legalSiren
    }

    return entrepriseNew
  })

export { entreprisesFormatTable }
