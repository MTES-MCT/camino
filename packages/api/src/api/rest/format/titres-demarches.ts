import { ITitreDemarche, Index } from '../../../types'

import { titreEtapesSortDescByOrdre } from '../../../business/utils/titre-etapes-sort'
import { getEtapesStatuts } from 'camino-common/src/static/etapesTypesEtapesStatuts'
import { DemarchesStatuts } from 'camino-common/src/static/demarchesStatuts'

const etapesDatesStatutsBuild = (titreDemarche: ITitreDemarche) => {
  if (!titreDemarche.etapes?.length) return null

  const etapes = titreEtapesSortDescByOrdre(titreDemarche.etapes)

  return etapes
    .filter(e => e.statutId !== 'aco')
    .reduce((etapesDatesStatuts, etape) => {
      const type = etape.type

      if (!type) return etapesDatesStatuts

      etapesDatesStatuts[type.nom] = etape?.date || ''

      // si le type d'Ã©tape a plus d'un statut possible
      // (ex : fav/def ou acc/rej)
      // alors on exporte aussi le statut

      const etapesStatuts = getEtapesStatuts(type.id)
      if (etapesStatuts!.length > 1) {
        const statut = etapesStatuts!.find(s => s.id === etape.statutId)

        etapesDatesStatuts[`${type.nom}_statut`] = statut?.nom || ''
      }

      return etapesDatesStatuts
    }, {} as Index<string>)
}

export const titresDemarchesFormatTable = (titresDemarches: ITitreDemarche[]) =>
  titresDemarches.map(titreDemarche => {
    const titre = titreDemarche.titre!

    const etapesTypesStatuts = etapesDatesStatutsBuild(titreDemarche)

    const etapeWithPoints = titreDemarche.etapes
      ? titreEtapesSortDescByOrdre(titreDemarche.etapes).find(
          etape => etape.points?.length
        )
      : undefined

    const titreDemarcheNew = {
      titre_id: titre.slug,
      titre_nom: titre.nom,
      titre_domaine: titre.domaine!.nom,
      titre_type: titre.type!.type.nom,
      titre_statut: titre.statut!.nom,
      type: titreDemarche.type!.nom,
      statut: DemarchesStatuts[titreDemarche.statutId!].nom,
      description: titreDemarche.description,
      titre_references: titre.references
        ?.map(r => `${r.type?.nom} : ${r.nom}`)
        .join(';'),
      titulaires_noms: titre.titulaires!.map(e => e.nom).join(';'),
      titulaires_adresses: titre
        .titulaires!.map(e => `${e.adresse} ${e.codePostal} ${e.commune}`)
        .join(';'),
      titulaires_legal: titre
        .titulaires!.map(e => e.legalEtranger || e.legalSiren)
        .join(';'),
      amodiataires_noms: titre.amodiataires!.map(e => e.nom).join(';'),
      amodiataires_adresses: titre
        .amodiataires!.map(e => `${e.adresse} ${e.codePostal} ${e.commune}`)
        .join(';'),
      amodiataires_legal: titre
        .amodiataires!.map(e => e.legalEtranger || e.legalSiren)
        .join(';'),
      ...etapesTypesStatuts,
      forets: etapeWithPoints
        ? etapeWithPoints.forets?.map(f => f.nom).join(';')
        : '',
      communes: etapeWithPoints
        ? etapeWithPoints.communes?.map(f => f.nom).join(';')
        : ''
    }

    return titreDemarcheNew
  })
