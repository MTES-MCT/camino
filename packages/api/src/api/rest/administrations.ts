import { Request as JWTRequest } from 'express-jwt'
import { HTTP_STATUS } from 'camino-common/src/http'

import { CustomResponse } from './express-type'
import { AdminUserNotNull, User, UserNotNull } from 'camino-common/src/roles'
import { Pool } from 'pg'
import { AdministrationId, administrationIdValidator } from 'camino-common/src/static/administrations'
import { canReadAdministrations } from 'camino-common/src/permissions/administrations'
import { deleteAdministrationActiviteTypeEmail, getActiviteTypeEmailsByAdministrationId, getUtilisateursByAdministrationId, insertAdministrationActiviteTypeEmail } from './administrations.queries'
import { AdministrationActiviteTypeEmail } from 'camino-common/src/administrations'
import { CaminoApiError } from '../../types'
import { ZodUnparseable } from '../../tools/fp-tools'
import { DeepReadonly } from 'camino-common/src/typescript-tools'
import { DbQueryAccessError } from '../../pg-database'
import { Effect, Match } from 'effect'
import { RestNewPostCall } from '../../server/rest'

export const getAdministrationUtilisateurs = (pool: Pool) => async (req: JWTRequest<User>, res: CustomResponse<AdminUserNotNull[]>) => {
  const user = req.auth

  const parsed = administrationIdValidator.safeParse(req.params.administrationId)

  if (!parsed.success) {
    console.warn(`l'administrationId est obligatoire`)
    res.sendStatus(HTTP_STATUS.FORBIDDEN)
  } else if (!canReadAdministrations(user)) {
    res.sendStatus(HTTP_STATUS.FORBIDDEN)
  } else {
    try {
      res.json(await getUtilisateursByAdministrationId(pool, parsed.data))
    } catch (e) {
      console.error(e)

      res.sendStatus(HTTP_STATUS.INTERNAL_SERVER_ERROR)
    }
  }
}

export const getAdministrationActiviteTypeEmails = (pool: Pool) => async (req: JWTRequest<User>, res: CustomResponse<AdministrationActiviteTypeEmail[]>) => {
  const user = req.auth

  const parsed = administrationIdValidator.safeParse(req.params.administrationId)

  if (!parsed.success) {
    console.warn(`l'administrationId est obligatoire`)
    res.sendStatus(HTTP_STATUS.FORBIDDEN)
  } else if (!canReadAdministrations(user)) {
    res.sendStatus(HTTP_STATUS.FORBIDDEN)
  } else {
    try {
      res.json(await getActiviteTypeEmailsByAdministrationId(pool, parsed.data))
    } catch (e) {
      console.error(e)

      res.sendStatus(HTTP_STATUS.INTERNAL_SERVER_ERROR)
    }
  }
}

export const addAdministrationActiviteTypeEmails: RestNewPostCall<'/rest/administrations/:administrationId/activiteTypeEmails'> = (
  pool: Pool,
  user: DeepReadonly<UserNotNull>,
  body: DeepReadonly<AdministrationActiviteTypeEmail>,
  params: { administrationId: AdministrationId }
): Effect.Effect<boolean, CaminoApiError<'Accès interdit' | ZodUnparseable | DbQueryAccessError>> => {
  return Effect.Do.pipe(
    Effect.filterOrFail(
      () => canReadAdministrations(user),
      () => ({ message: 'Accès interdit' as const })
    ),
    Effect.flatMap(() => insertAdministrationActiviteTypeEmail(pool, params.administrationId, body)),
    Effect.mapError(caminoError =>
      Match.value(caminoError.message).pipe(
        Match.when('Accès interdit', () => ({ ...caminoError, status: HTTP_STATUS.FORBIDDEN })),
        Match.when('Problème de validation de données', () => ({ ...caminoError, status: HTTP_STATUS.BAD_REQUEST })),
        Match.when("Impossible d'accéder à la base de données", () => ({ ...caminoError, status: HTTP_STATUS.INTERNAL_SERVER_ERROR })),
        Match.exhaustive
      )
    )
  )
}

export const deleteAdministrationActiviteTypeEmails: RestNewPostCall<'/rest/administrations/:administrationId/activiteTypeEmails/delete'> = (
  pool: Pool,
  user: DeepReadonly<UserNotNull>,
  body: DeepReadonly<AdministrationActiviteTypeEmail>,
  params: { administrationId: AdministrationId }
): Effect.Effect<boolean, CaminoApiError<'Accès interdit' | ZodUnparseable | DbQueryAccessError>> => {
  return Effect.Do.pipe(
    Effect.filterOrFail(
      () => canReadAdministrations(user),
      () => ({ message: 'Accès interdit' as const })
    ),
    Effect.flatMap(() => deleteAdministrationActiviteTypeEmail(pool, params.administrationId, body)),
    Effect.mapError(caminoError =>
      Match.value(caminoError.message).pipe(
        Match.when('Accès interdit', () => ({ ...caminoError, status: HTTP_STATUS.FORBIDDEN })),
        Match.when('Problème de validation de données', () => ({ ...caminoError, status: HTTP_STATUS.BAD_REQUEST })),
        Match.when("Impossible d'accéder à la base de données", () => ({ ...caminoError, status: HTTP_STATUS.INTERNAL_SERVER_ERROR })),
        Match.exhaustive
      )
    )
  )
}
