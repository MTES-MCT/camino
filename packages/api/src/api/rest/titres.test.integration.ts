import { dbManager } from '../../../tests/db-manager.js'
import { titreCreate, titreUpdate } from '../../database/queries/titres.js'
import { titreDemarcheCreate } from '../../database/queries/titres-demarches.js'
import { titreEtapeCreate } from '../../database/queries/titres-etapes.js'
import { userSuper } from '../../database/user-super.js'
import { restCall, restDeleteCall, restPostCall } from '../../../tests/_utils/index.js'
import { ADMINISTRATION_IDS } from 'camino-common/src/static/administrations.js'
import { ITitreDemarche, ITitreEtape } from '../../types.js'
import { entreprisesUpsert } from '../../database/queries/entreprises.js'
import { Knex } from 'knex'
import { toCaminoDate } from 'camino-common/src/date.js'
import { afterAll, beforeAll, beforeEach, describe, test, expect, vi } from 'vitest'
import { newEntrepriseId } from 'camino-common/src/entreprise.js'
import type { Pool } from 'pg'
import { newDemarcheId, newEtapeId, newTitreId } from '../../database/models/_format/id-create.js'
import { HTTP_STATUS } from 'camino-common/src/http.js'
import { toCommuneId } from 'camino-common/src/static/communes.js'
import { insertCommune } from '../../database/queries/communes.queries.js'
import { titreSlugValidator } from 'camino-common/src/validators/titres.js'
import TitresDemarches from '../../database/models/titres-demarches.js'
import TitresEtapes from '../../database/models/titres-etapes.js'
import Titres from '../../database/models/titres.js'
import { ETAPE_IS_NOT_BROUILLON } from 'camino-common/src/etape.js'

console.info = vi.fn()
console.error = vi.fn()

let knex: Knex<any, unknown[]>
let dbPool: Pool
beforeAll(async () => {
  const { knex: knexInstance, pool } = await dbManager.populateDb()
  dbPool = pool
  knex = knexInstance

  await insertCommune(pool, { id: toCommuneId('97300'), nom: 'Une ville en Guyane' })
  const entreprises = await entreprisesUpsert([{ id: newEntrepriseId('plop'), nom: 'Mon Entreprise' }])
  await titreCreate(
    {
      nom: 'mon titre simple',
      typeId: 'arm',
      titreStatutId: 'val',
      propsTitreEtapesIds: {},
    },
    {}
  )

  await createTitreWithEtapes(
    'titre1',
    [
      {
        typeId: 'mfr',
        statutId: 'fai',
        isBrouillon: ETAPE_IS_NOT_BROUILLON,
        date: toCaminoDate('2022-01-01'),
        ordre: 0,
        administrationsLocales: [ADMINISTRATION_IDS['DGTM - GUYANE']],
        communes: [{ id: toCommuneId('97300'), surface: 12 }],
      },
      {
        typeId: 'mdp',
        statutId: 'fai',
        isBrouillon: ETAPE_IS_NOT_BROUILLON,
        date: toCaminoDate('2022-02-01'),
        ordre: 1,
      },
      {
        typeId: 'pfd',
        statutId: 'fai',
        isBrouillon: ETAPE_IS_NOT_BROUILLON,
        date: toCaminoDate('2022-02-10'),
        ordre: 2,
      },
      {
        typeId: 'mcp',
        statutId: 'com',
        isBrouillon: ETAPE_IS_NOT_BROUILLON,
        date: toCaminoDate('2022-03-10'),
        ordre: 3,
      },
    ],
    entreprises
  )
  await createTitreWithEtapes(
    'titre2',
    [
      {
        typeId: 'mfr',
        statutId: 'fai',
        isBrouillon: ETAPE_IS_NOT_BROUILLON,
        date: toCaminoDate('2022-01-01'),
        ordre: 0,
        administrationsLocales: [ADMINISTRATION_IDS['DGTM - GUYANE']],
      },
      {
        typeId: 'mdp',
        statutId: 'fai',
        isBrouillon: ETAPE_IS_NOT_BROUILLON,
        date: toCaminoDate('2022-02-01'),
        ordre: 1,
      },
      {
        typeId: 'pfd',
        statutId: 'fai',
        isBrouillon: ETAPE_IS_NOT_BROUILLON,
        date: toCaminoDate('2022-02-10'),
        ordre: 2,
      },
    ],
    entreprises
  )
})

afterAll(async () => {
  await dbManager.closeKnex()
})

const titreEtapesCreate = async (demarche: ITitreDemarche, etapes: Omit<ITitreEtape, 'id' | 'titreDemarcheId'>[]) => {
  const etapesCrees = []
  for (const etape of etapes) {
    etapesCrees.push(
      await titreEtapeCreate(
        {
          ...etape,
          titreDemarcheId: demarche.id,
        },
        userSuper,
        demarche.titreId
      )
    )
  }

  return etapesCrees
}

async function createTitreWithEtapes(nomTitre: string, etapes: Omit<ITitreEtape, 'id' | 'titreDemarcheId'>[], entreprises: any) {
  const titre = await titreCreate(
    {
      nom: nomTitre,
      typeId: 'arm',
      titreStatutId: 'mod',
      propsTitreEtapesIds: {},
      references: [
        {
          referenceTypeId: 'onf',
          nom: 'ONF',
        },
      ],
    },
    {}
  )

  const titreDemarche = await titreDemarcheCreate({
    titreId: titre.id,
    typeId: 'oct',
  })

  const [firstEtape, ...remainingEtapes] = etapes

  const etapesCrees = await titreEtapesCreate(titreDemarche, [{ ...firstEtape, titulaireIds: [entreprises[0].id] }, ...remainingEtapes])

  await knex('titres')
    .update({ propsTitreEtapesIds: { titulaires: etapesCrees[0].id, points: etapesCrees[0].id } })
    .where('id', titre.id)

  return titre.id
}

describe('titresONF', () => {
  test("teste la récupération des données pour l'ONF", async () => {
    const tested = await restCall(
      dbPool,
      '/rest/titresONF',
      {},
      {
        role: 'admin',
        administrationId: ADMINISTRATION_IDS['OFFICE NATIONAL DES FORÊTS'],
      }
    )

    expect(tested.statusCode).toBe(200)
    expect(tested.body).toHaveLength(2)
    expect(tested.body[0]).toMatchSnapshot({
      id: expect.any(String),
      slug: expect.any(String),
    })
    expect(tested.body[1]).toMatchSnapshot({
      id: expect.any(String),
      slug: expect.any(String),
    })
  })
})

describe('titresAdministration', () => {
  test('teste la récupération des données pour les Administrations', async () => {
    const tested = await restCall(
      dbPool,
      '/rest/titresAdministrations',
      {},
      {
        role: 'admin',
        administrationId: ADMINISTRATION_IDS['DGTM - GUYANE'],
      }
    )

    expect(tested.statusCode).toBe(200)
    expect(tested.body).toHaveLength(2)
    expect(tested.body[0]).toMatchSnapshot({
      id: expect.any(String),
      slug: expect.any(String),
    })
    expect(tested.body[1]).toMatchSnapshot({
      id: expect.any(String),
      slug: expect.any(String),
    })
  })
})
describe('titresLiaisons', () => {
  test('peut lier deux titres', async () => {
    const getTitres = await restCall(
      dbPool,
      '/rest/titresONF',
      {},
      {
        role: 'admin',
        administrationId: ADMINISTRATION_IDS['OFFICE NATIONAL DES FORÊTS'],
      }
    )
    const titreId = getTitres.body[0].id

    const axm = await titreCreate(
      {
        nom: 'mon axm simple',
        typeId: 'axm',
        titreStatutId: 'val',
        propsTitreEtapesIds: {},
      },
      {}
    )

    const tested = await restPostCall(
      dbPool,
      '/rest/titres/:id/titreLiaisons',
      { id: axm.id },
      {
        role: 'admin',
        administrationId: ADMINISTRATION_IDS['OFFICE NATIONAL DES FORÊTS'],
      },
      [titreId]
    )

    expect(tested.statusCode).toBe(200)
    expect(tested.body.amont).toHaveLength(1)
    expect(tested.body.aval).toHaveLength(0)
    expect(tested.body.amont[0]).toStrictEqual({
      id: titreId,
      nom: getTitres.body[0].nom,
    })

    const avalTested = await restCall(
      dbPool,
      '/rest/titres/:id/titreLiaisons',
      { id: titreId },
      {
        role: 'admin',
        administrationId: ADMINISTRATION_IDS['OFFICE NATIONAL DES FORÊTS'],
      }
    )

    expect(avalTested.statusCode).toBe(200)
    expect(avalTested.body.amont).toHaveLength(0)
    expect(avalTested.body.aval).toHaveLength(1)
    expect(avalTested.body.aval[0]).toStrictEqual({
      id: axm.id,
      nom: axm.nom,
    })
  })
})

describe('titreModifier', () => {
  let id = newTitreId('')

  beforeEach(async () => {
    const titre = await titreCreate(
      {
        nom: 'mon titre',
        typeId: 'arm',
        titreStatutId: 'ind',
        propsTitreEtapesIds: {},
      },
      {}
    )
    id = titre.id
  })

  test('ne peut pas modifier un titre (utilisateur anonyme)', async () => {
    const tested = await restPostCall(dbPool, '/rest/titres/:titreId', { titreId: id }, undefined, { id, nom: 'mon titre modifié', references: [] })

    expect(tested.statusCode).toBe(403)
  })

  test("ne peut pas modifier un titre (un utilisateur 'entreprise')", async () => {
    const tested = await restPostCall(dbPool, '/rest/titres/:titreId', { titreId: id }, { role: 'entreprise', entreprises: [] }, { id, nom: 'mon titre modifié', references: [] })

    expect(tested.statusCode).toBe(404)
  })

  test('modifie un titre (un utilisateur userSuper)', async () => {
    const tested = await restPostCall(dbPool, '/rest/titres/:titreId', { titreId: id }, { role: 'super' }, { id, nom: 'mon titre modifié', references: [] })
    expect(tested.statusCode).toBe(204)
  })

  test("modifie un titre ARM (un utilisateur 'admin' PTMG)", async () => {
    const tested = await restPostCall(
      dbPool,
      '/rest/titres/:titreId',
      { titreId: id },
      {
        role: 'admin',
        administrationId: ADMINISTRATION_IDS['PÔLE TECHNIQUE MINIER DE GUYANE'],
      },
      { id, nom: 'mon titre modifié', references: [] }
    )
    expect(tested.statusCode).toBe(204)
  })

  test("ne peut pas modifier un titre ARM échu (un utilisateur 'admin' PTMG)", async () => {
    const titre = await titreCreate(
      {
        nom: 'mon titre échu',
        typeId: 'arm',
        titreStatutId: 'ech',
        propsTitreEtapesIds: {},
      },
      {}
    )

    const tested = await restPostCall(
      dbPool,
      '/rest/titres/:titreId',
      { titreId: titre.id },
      {
        role: 'admin',
        administrationId: ADMINISTRATION_IDS['PÔLE TECHNIQUE MINIER DE GUYANE'],
      },
      { id: titre.id, nom: 'mon titre modifié', references: [] }
    )
    expect(tested.statusCode).toBe(403)
  })

  test("ne peut pas modifier un titre ARM (un utilisateur 'admin' DGCL/SDFLAE/FL1)", async () => {
    const tested = await restPostCall(
      dbPool,
      '/rest/titres/:titreId',
      { titreId: id },
      { role: 'admin', administrationId: ADMINISTRATION_IDS['DGCL/SDFLAE/FL1'] },
      { id, nom: 'mon titre modifié', references: [] }
    )
    expect(tested.statusCode).toBe(403)
  })
})

describe('titreSupprimer', () => {
  let id = newTitreId('')

  beforeEach(async () => {
    const titre = await titreCreate(
      {
        nom: 'mon titre',
        typeId: 'arm',
        titreStatutId: 'ind',
        propsTitreEtapesIds: {},
      },
      {}
    )
    id = titre.id
  })

  test('ne peut pas supprimer un titre (utilisateur anonyme)', async () => {
    const tested = await restDeleteCall(dbPool, '/rest/titres/:titreId', { titreId: id }, undefined)

    expect(tested.statusCode).toBe(403)
  })

  test('peut supprimer un titre (utilisateur super)', async () => {
    const tested = await restDeleteCall(dbPool, '/rest/titres/:titreId', { titreId: id }, userSuper)

    expect(tested.statusCode).toBe(204)
  })

  test('ne peut pas supprimer un titre inexistant (utilisateur super)', async () => {
    const tested = await restDeleteCall(dbPool, '/rest/titres/:titreId', { titreId: newTitreId('toto') }, userSuper)
    expect(tested.statusCode).toBe(404)
  })
})

describe('getTitre', () => {
  test('ne peut pas récupérer un titre inexistant', async () => {
    const tested = await restCall(dbPool, '/rest/titres/:titreId', { titreId: newTitreId('not existing') }, userSuper)

    expect(tested.statusCode).toBe(404)
  })

  test('peut récupérer un titre', async () => {
    const titreId = newTitreId('other-titre-id')
    await Titres.query().insert({
      id: titreId,
      nom: 'mon nouveau titre',
      typeId: 'arm',
      titreStatutId: 'ind',
      slug: titreSlugValidator.parse('arm-slug'),
      propsTitreEtapesIds: {},
      archive: false,
    })

    const tested = await restCall(dbPool, '/rest/titres/:titreId', { titreId }, userSuper)

    expect(tested.statusCode).toBe(200)
    expect(tested.body).toMatchInlineSnapshot(`
      {
        "demarches": [],
        "id": "other-titre-id",
        "nb_activites_to_do": null,
        "nom": "mon nouveau titre",
        "references": [],
        "slug": "arm-slug",
        "titre_doublon": null,
        "titre_last_modified_date": null,
        "titre_statut_id": "ind",
        "titre_type_id": "arm",
      }
    `)
  })
  test('ne peut pas récupérer un titre archivé', async () => {
    const titreId = newTitreId('archive-titre-id')
    await Titres.query().insert({
      id: titreId,
      nom: 'mon nouveau titre',
      typeId: 'arm',
      titreStatutId: 'ind',
      slug: titreSlugValidator.parse('arm-slug'),
      propsTitreEtapesIds: {},
      archive: true,
    })

    const tested = await restCall(dbPool, '/rest/titres/:titreId', { titreId }, userSuper)

    expect(tested.statusCode).toBe(404)
    expect(tested.body).toMatchInlineSnapshot(`{}`)
  })

  test('peut récupérer un titre avec des administrations locales', async () => {
    const titreId = newTitreId('titre-id')
    await Titres.query().insert({
      id: titreId,
      nom: 'mon titre',
      typeId: 'arm',
      slug: titreSlugValidator.parse('slug'),
      titreStatutId: 'val',
      propsTitreEtapesIds: {},
      archive: false,
    })

    const demarcheId = newDemarcheId('demarche-id')
    await TitresDemarches.query().insert({
      id: demarcheId,
      titreId,
      typeId: 'oct',
    })

    const etapeId = newEtapeId('titre-etape-id')
    await TitresEtapes.query().insert({
      id: etapeId,
      titreDemarcheId: demarcheId,
      typeId: 'mfr',
      statutId: 'fai',
      date: toCaminoDate('2022-01-01'),
      ordre: 0,
      administrationsLocales: ['aut-97300-01', 'aut-mrae-guyane-01'],
    })

    await titreUpdate(titreId, { propsTitreEtapesIds: { points: etapeId } })

    const tested = await restCall(dbPool, '/rest/titres/:titreId', { titreId }, userSuper)

    expect(tested.statusCode).toBe(200)
    expect(tested.body).toMatchInlineSnapshot(`
      {
        "demarches": [
          {
            "demarche_date_debut": null,
            "demarche_date_fin": null,
            "demarche_statut_id": "ind",
            "demarche_type_id": "oct",
            "description": null,
            "etapes": [
              {
                "avis_documents": [],
                "date": "2022-01-01",
                "entreprises_documents": [],
                "etape_documents": [],
                "etape_statut_id": "fai",
                "etape_type_id": "mfr",
                "fondamentale": {
                  "amodiataireIds": null,
                  "date_debut": null,
                  "date_fin": null,
                  "duree": null,
                  "perimetre": null,
                  "substances": [],
                  "titulaireIds": null,
                },
                "id": "titre-etape-id",
                "is_brouillon": false,
                "note": {
                  "is_avertissement": false,
                  "valeur": "",
                },
                "ordre": 0,
                "sections_with_values": [
                  {
                    "elements": [
                      {
                        "description": "",
                        "id": "mecanise",
                        "nom": "Prospection mécanisée",
                        "optionnel": false,
                        "type": "radio",
                        "value": null,
                      },
                      {
                        "description": "Nombre de franchissements de cours d'eau",
                        "id": "franchissements",
                        "nom": "Franchissements de cours d'eau",
                        "optionnel": true,
                        "type": "integer",
                        "value": null,
                      },
                    ],
                    "id": "arm",
                    "nom": "Caractéristiques ARM",
                  },
                ],
                "slug": "demarche-id-mfr99",
              },
            ],
            "id": "demarche-id",
            "ordre": 0,
            "slug": "titre-id-oct99",
          },
        ],
        "id": "titre-id",
        "nb_activites_to_do": null,
        "nom": "mon titre",
        "references": [],
        "slug": "slug",
        "titre_doublon": null,
        "titre_last_modified_date": null,
        "titre_statut_id": "val",
        "titre_type_id": "arm",
      }
    `)
  })
})

test('utilisateurTitreAbonner', async () => {
  const titre = await titreCreate(
    {
      nom: 'mon autre titre',
      typeId: 'arm',
      slug: titreSlugValidator.parse('slug'),
      titreStatutId: 'val',
      propsTitreEtapesIds: {},
    },
    {}
  )

  const tested = await restPostCall(dbPool, '/rest/titres/:titreId/abonne', { titreId: titre.id }, userSuper, { abonne: true })

  expect(tested.statusCode).toBe(HTTP_STATUS.HTTP_STATUS_NO_CONTENT)
})
