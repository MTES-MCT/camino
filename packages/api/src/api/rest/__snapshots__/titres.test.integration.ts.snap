// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`titresDREAL > teste la récupération des données pour les DREAL 1`] = `
{
  "activitesAbsentes": 0,
  "activitesEnConstruction": 0,
  "derniereEtape": {
    "date": "2022-03-10",
    "etapeStatutId": "com",
    "etapeTypeId": "mcp",
  },
  "enAttenteDeDREAL": false,
  "id": Any<String>,
  "nom": "titre1",
  "prochainesEtapes": [
    "vfd",
    "mod",
  ],
  "references": [
    {
      "nom": "ONF",
      "referenceTypeId": "onf",
    },
  ],
  "slug": Any<String>,
  "titre_statut_id": "mod",
  "titulaires": [
    {
      "adresse": null,
      "archive": false,
      "categorie": null,
      "cedex": null,
      "codePostal": null,
      "commune": null,
      "dateCreation": null,
      "email": null,
      "id": "plop",
      "legalEtranger": null,
      "legalForme": null,
      "legalSiren": null,
      "nom": "Mon Entreprise",
      "operateur": null,
      "paysId": null,
      "telephone": null,
      "url": null,
    },
  ],
  "type_id": "arm",
}
`;

exports[`titresDREAL > teste la récupération des données pour les DREAL 2`] = `
{
  "activitesAbsentes": 0,
  "activitesEnConstruction": 0,
  "derniereEtape": {
    "date": "2022-02-10",
    "etapeStatutId": "fai",
    "etapeTypeId": "pfd",
  },
  "enAttenteDeDREAL": false,
  "id": Any<String>,
  "nom": "titre2",
  "prochainesEtapes": [
    "mcp",
    "mod",
  ],
  "references": [
    {
      "nom": "ONF",
      "referenceTypeId": "onf",
    },
  ],
  "slug": Any<String>,
  "titre_statut_id": "mod",
  "titulaires": [
    {
      "adresse": null,
      "archive": false,
      "categorie": null,
      "cedex": null,
      "codePostal": null,
      "commune": null,
      "dateCreation": null,
      "email": null,
      "id": "plop",
      "legalEtranger": null,
      "legalForme": null,
      "legalSiren": null,
      "nom": "Mon Entreprise",
      "operateur": null,
      "paysId": null,
      "telephone": null,
      "url": null,
    },
  ],
  "type_id": "arm",
}
`;

exports[`titresONF > teste la récupération des données pour l'ONF 1`] = `
{
  "dateCARM": "",
  "dateCompletudePTMG": "2022-03-10",
  "dateReceptionONF": "",
  "enAttenteDeONF": true,
  "id": Any<String>,
  "nom": "titre1",
  "references": [
    {
      "nom": "ONF",
      "referenceTypeId": "onf",
    },
  ],
  "slug": Any<String>,
  "titre_statut_id": "mod",
  "titulaires": [
    {
      "nom": "Mon Entreprise",
    },
  ],
  "type_id": "arm",
}
`;

exports[`titresONF > teste la récupération des données pour l'ONF 2`] = `
{
  "dateCARM": "",
  "dateCompletudePTMG": "",
  "dateReceptionONF": "",
  "enAttenteDeONF": false,
  "id": Any<String>,
  "nom": "titre2",
  "references": [
    {
      "nom": "ONF",
      "referenceTypeId": "onf",
    },
  ],
  "slug": Any<String>,
  "titre_statut_id": "mod",
  "titulaires": [
    {
      "nom": "Mon Entreprise",
    },
  ],
  "type_id": "arm",
}
`;

exports[`titresPTMG > teste la récupération des données pour le PTMG 1`] = `
{
  "enAttenteDePTMG": false,
  "id": Any<String>,
  "nom": "titre1",
  "references": [
    {
      "nom": "ONF",
      "referenceTypeId": "onf",
    },
  ],
  "slug": Any<String>,
  "titre_statut_id": "mod",
  "titulaires": [
    {
      "nom": "Mon Entreprise",
    },
  ],
  "type_id": "arm",
}
`;

exports[`titresPTMG > teste la récupération des données pour le PTMG 2`] = `
{
  "enAttenteDePTMG": true,
  "id": Any<String>,
  "nom": "titre2",
  "references": [
    {
      "nom": "ONF",
      "referenceTypeId": "onf",
    },
  ],
  "slug": Any<String>,
  "titre_statut_id": "mod",
  "titulaires": [
    {
      "nom": "Mon Entreprise",
    },
  ],
  "type_id": "arm",
}
`;
